src/Umbraco.Core/Services/DomainService.cs

```diff
@@ -65,15 +65,15 @@ public IDomain GetByName(string name)
             }
         }
 
-        public IEnumerable<IDomain> GetAll(bool includeWildcards)
+        public IEnumerable<IDomain>? GetAll(bool includeWildcards)
```

Comment:
As discussed, would likely be better here (and any other instances) to ensure that collections are _not_ nullable - so the caller doesn't have to null check and we align with the best practice of having collections return empty rather than null when there is no data.

Reply:
Agreed, unless there's really a syntactic difference between having no value and no results.

This is also mentioned in the [MS design guidelines](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/guidelines-for-collections#collection-properties-and-return-values):

> DO NOT return null values from collection properties or from methods returning collections. Return an empty collection or an empty array instead.

One downside is that this would require iterating the returned items to see if there are any (although only doing a null check won't guarantee there are items either). Maybe it would be better to return a collection type like `ICollection<T>` or `IReadOnlyCollection<T>`, so you can easily check the amount of items returned (and ensure iterating it doesn't have any side effects, like deferred or multiple execution)...

Reply:
`[This](https://enterprisecraftsmanship.com/posts/ienumerable-vs-ireadonlylist/)` is worth a read on the topic, (TLDR: prefer `IReadOnlyList` for returning collections).

Reply:
Thank you for raising this @AndyButland, this is now fixed in https://github.com/umbraco/Umbraco-CMS/pull/12221 🚀

Reply:
@AndyButland A good read indeed and changing `IEnumerable<T>` to `IReadOnlyList<T>` would be the best conservative return type yes 👍🏻

And to prevent casting the original list back to `IList<T>`, you could wrap it inside an [`ReadOnlyCollection<T>` using `AsReadOnly()`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.asreadonly).

Reply:
@Zeegaan The same goes for all other places where an `IEnumerable<T>?` is currently returned, but shouldn't (e.g. if a null value would return an exception or by falling back to `Enumerable.Empty<T>()`).